src/native/managed/cdacreader/Microsoft.Diagnostics.DataContractReader.Contracts/Contracts/ExecutionManager/Helpers/NibbleMapHelpers.cs

```diff
@@ -0,0 +1,147 @@
+        public MapUnit ShiftNextNibble => new MapUnit(Value >>> 4);
+
+        public const uint NibbleMask = 0x0Fu;
+        internal Nibble Nibble => new((int)(Value & NibbleMask));
+
+        public bool IsZero => Value == 0;
+
+        // Assuming mapIdx is the index of a nibble within the current map unit,
+        // shift the unit so that nibble is in the least significant position and return the result.
+        public MapUnit FocusOnIndexedNibble(MapKey mapIdx)
+        {
+            int shift = mapIdx.GetNibbleShift();
+            return new MapUnit (Value >>> shift);
+        }
+    }
+
+    // Each nibble is a 4-bit integer that gives an offset within a bucket.
+    // We reserse 0 to mean that there is no method starting at any offset within a bucket
+    internal readonly struct Nibble
+    {
+        public readonly int Value;
+
+        public Nibble(int value)
+        {
+            Debug.Assert (value >= 0 && value <= 0xF);
+            Value = value;
+        }
+
+        public static Nibble Zero => new Nibble(0);
+        public bool IsEmpty => Value == 0;
+
+        public ulong TargetByteOffset
+        {
+            get
+            {
+                Debug.Assert(Value != 0);
+                return (uint)(Value - 1) * MapUnit.SizeInBytes;
+            }
+        }
+    }
+
+    // The key to the map is the index of an individual nibble
+    internal readonly struct MapKey
+    {
+        private readonly ulong MapIdx;
+        public MapKey(ulong mapIdx) => MapIdx = mapIdx;
+        public override string ToString() => $"0x{MapIdx:x}";
+
+        // The offset of the address in the target space that this map index represents
+        public ulong TargetByteOffset => MapIdx * BytesPerBucket;
+
+        // The index of the map unit that contains this map index
+        public ulong ContainingMapUnitIndex => MapIdx / MapUnit.SizeInNibbles;
+
+        // The offset of the map unit that contains this map index
+        public ulong ContainingMapUnitByteOffset => ContainingMapUnitIndex * MapUnit.SizeInBytes;
+
+        // The map index is the index of a nibble within the map, this gives the index of that nibble within a map unit.
+        public int NibbleIndexInMapUnit => (int)(MapIdx & (MapUnit.SizeInNibbles - 1));
+
+        // go to the previous nibble
+        public MapKey Prev => new MapKey(MapIdx - 1);
+
+        // to to the previous map unit
+        public MapKey PrevMapUnit => new MapKey(MapIdx - MapUnit.SizeInNibbles);
+
+        // Get a MapKey that is aligned to the first nibble in the map unit that contains this map index
+        public MapKey AlignDownToMapUnit() =>new MapKey(MapIdx & (~(MapUnit.SizeInNibbles - 1)));
+
+        // If the map index is less than the size of a map unit, we are in the first MapUnit and
+        // can stop searching
+        public bool InFirstMapUnit => MapIdx < MapUnit.SizeInNibbles;
+
+        public bool IsZero => MapIdx == 0;
```

Comment:
Align with the other data structures.

Reply:
I think this one should be different than the `IsEmpty` on the `Nibble` or `MapUnit`. Both of those can be "empty", this checks if we are at index zero in the map.